src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchEntityMapper.java

```diff
@@ -0,0 +1,676 @@
+	}
+
+	// --> READ
+
+	@Override
+	public <T> T readObject(Map<String, Object> source, Class<T> targetType) {
+		return read(targetType, source);
+	}
+
+	@Override
+	@Nullable
+	public <R> R read(Class<R> type, Map source) {
+		return doRead(source, ClassTypeInformation.from((Class<R>) ClassUtils.getUserClass(type)));
+	}
+
+	@Nullable
+	protected <R> R doRead(Map source, TypeInformation<R> typeHint) {
+
+		if (source == null) {
+			return null;
+		}
+
+
+
+		if(typeHint.getType().isInterface()) {
+			typeHint = typeMapper.readType(source, typeHint);
+		}
+
+		if (conversions.hasCustomReadTarget(Map.class, typeHint.getType())) {
+			return conversionService.convert(source, typeHint.getType());
+		}
+
+		if (typeHint.isMap() || ClassTypeInformation.OBJECT.equals(typeHint)) {
+			return (R) source;
+		}
+
+		ElasticsearchPersistentEntity<?> entity = mappingContext.getRequiredPersistentEntity(typeHint);
+		return readEntity(entity, source);
+	}
+
+	protected <R> R readEntity(ElasticsearchPersistentEntity<?> entity, Map source) {
+
+		ElasticsearchPersistentEntity<?> targetEntity = computeClosestEntity(entity, source);
+
+		ElasticsearchPropertyValueProvider propertyValueProvider = new ElasticsearchPropertyValueProvider(
+				new MapValueAccessor(source));
+
+		EntityInstantiator instantiator = instantiators.getInstantiatorFor(targetEntity);
+
+		R instance = (R) instantiator.createInstance(targetEntity,
+				new PersistentEntityParameterValueProvider<>(targetEntity, propertyValueProvider, null));
+
+		if (targetEntity.requiresPropertyPopulation()) {
+			return readProperties(targetEntity, instance, propertyValueProvider);
+		}
+
+		return instance;
+	}
+
+	protected <R> R readProperties(ElasticsearchPersistentEntity<?> entity, R instance,
+			ElasticsearchPropertyValueProvider valueProvider) {
+
+		PersistentPropertyAccessor<R> accessor = new ConvertingPropertyAccessor<>(entity.getPropertyAccessor(instance),
+				conversionService);
+
+		for (ElasticsearchPersistentProperty prop : entity) {
+
+			if (entity.isConstructorArgument(prop) || prop.isScoreProperty()) {
+				continue;
+			}
+
+			Object value = valueProvider.getPropertyValue(prop);
+			if (value != null) {
+				accessor.setProperty(prop, valueProvider.getPropertyValue(prop));
+			}
+		}
+
+		return accessor.getBean();
+	}
+
+	protected <R> R readValue(@Nullable Object source, ElasticsearchPersistentProperty property,
+			TypeInformation<R> targetType) {
+
+		if (source == null) {
+			return null;
+		}
+
+		Class<?> rawType = targetType.getType();
+		if (conversions.hasCustomReadTarget(source.getClass(), rawType)) {
+			return (R) conversionService.convert(source, rawType);
+		} else if (source instanceof List) {
+			return readCollectionValue((List) source, property, targetType);
+		} else if (source instanceof Map) {
+			return readMapValue((Map<String, Object>) source, property, targetType);
+		} else if (Enum.class.isAssignableFrom(rawType)) {
```

Comment:
The code here is rather source-specific than target-type specific. Do we want to move this into `readSimpleValue(…)`?